### PR TITLE
Fix finalizer_ref usage of lock/wait resulting in task switches in gc context

### PR DESCRIPTION
**Pull request by krynju**
Restructured some code, so that lock on the `client_refs` is obtained once at the beginning of the finalizer.
It's obtained using `trylock` instead of `lock` down the line in the functions invoked later.
Usage of `lock` in gc context led to calling `wait()` and a task switch attempt.
Catching the error and trying again led to different errors (switching to exited tasks) due to failed task switch handling in `ensure_scheduled` in `task.jl`, so I went for repeated `trylock` instead. 

Fix addresses the two following issues found in https://github.com/JuliaParallel/Dagger.jl/issues/243

And supersedes https://github.com/JuliaLang/julia/pull/41786

```
error in running finalizer: ErrorException("task switch not allowed from inside gc finalizer")
jl_error at /opt/julia/src/rtutils.c:41
jl_switch at /opt/julia/src/task.c:481
try_yieldto at ./task.jl:754
wait at ./task.jl:824
wait at ./condition.jl:112
lock at ./lock.jl:100
lock at ./lock.jl:185
lock at ./weakkeydict.jl:89 [inlined]
delete! at ./weakkeydict.jl:168 [inlined]
finalize_ref at /opt/julia/usr/share/julia/stdlib/v1.8/Distributed/src/remotecall.jl:92
jl_apply at /opt/julia/src/julia.h:1768 [inlined]
run_finalizer at /opt/julia/src/gc.c:278
```
```
error in running finalizer: ErrorException("task switch not allowed from inside gc finalizer")
jl_error at /tmp/julia/src/rtutils.c:41
jl_switch at /tmp/julia/src/task.c:481
try_yieldto at ./task.jl:754
wait at ./task.jl:824
wait at ./condition.jl:112
lock at ./lock.jl:100
lock at ./lock.jl:185
lock at ./weakkeydict.jl:90 [inlined]
del_client at /tmp/julia/usr/share/julia/stdlib/v1.8/Distributed/src/remotecall.jl:236 [inlined]
del_client at /tmp/julia/usr/share/julia/stdlib/v1.8/Distributed/src/remotecall.jl:234 [inlined]
del_client at /tmp/julia/usr/share/julia/stdlib/v1.8/Distributed/src/remotecall.jl:232 [inlined]
send_del_client at /tmp/julia/usr/share/julia/stdlib/v1.8/Distributed/src/remotecall.jl:265
finalize_ref at /tmp/julia/usr/share/julia/stdlib/v1.8/Distributed/src/remotecall.jl:102
jl_apply at /tmp/julia/src/julia.h:1768 [inlined]
run_finalizer at /tmp/julia/src/gc.c:278
```

### CODE REVIEW

File: stdlib/Distributed/src/remotecall.jl [hunk 1]
```diff
@@ -229,18 +237,22 @@ del_client(rr::AbstractRemoteRef) = del_client(remoteref_id(rr), myid())
...
+function _del_client(pg, id, client)
+    rv = get(pg.refs, id, false)
+    if rv !== false
+        delete!(rv.clientset, client)
+        if isempty(rv.clientset)
+            delete!(pg.refs, id)
+            #print("$(myid()) collected $id\n")
+        end
+    end
+end
```
Comment (vtjnash) [inline stdlib/Distributed/src/remotecall.jl:254]: returning pg.refs seems odd. add a return `nothing`?

File: stdlib/Distributed/src/remotecall.jl [hunk 2]
```diff
@@ -84,20 +84,28 @@ end
...
+                    r.v = nothing
+                end
+                r.where = 0
+
+            catch
+                unlock(client_refs.lock)
+                finalizer(finalize_ref, r)
+                return nothing
+            finally
+                unlock(client_refs.lock)
```
Comment (vtjnash) [inline stdlib/Distributed/src/remotecall.jl:105]: this is a double-unlock (both error and return cases)

File: stdlib/Distributed/src/remotecall.jl [hunk 3]
```diff
@@ -84,20 +84,28 @@ end
...
+            end
         end
```
Comment (vtjnash) [inline stdlib/Distributed/src/remotecall.jl:107]:
```suggestion
        else
            # delay finalizer for later, when it's not already locked
            finalizer(finalize_ref, r)
            return nothing
        end
```

**Comment by krynju** [comment]
@vtjnash Thanks for the review! Forgot to cleanup that trylock logic, but I made it nice now.
Tested on two separate examples that were causing the issues mentioned in the main post and it's working well.

**Review by vtjnash — APPROVED** [review]

**Comment by krynju** [comment]
@Sacha0 Hey, could this be backported to 1.7? It's a relatively common error occuring in Dagger under any heavier usage. Would be awesome if this could be already fixed on 1.7 when it releases!

**Comment by Sacha0** [comment]
Tagging for backport --- thanks for the note Krystian! :)

**Pull request merged**